### ISSUE TITLE
Fine-tuning the speechbrain/asr-transformer-transformerlm-librispeech pretrained model

### ISSUE THREAD

**Issue by jclarke98** [speechbrain/speechbrain#1230]
Hi there,
I am trying to fine-tune the pre-trained transformer ASR model - [asr-transformer-transformerlm-librispeech](https://huggingface.co/speechbrain/asr-transformer-transformerlm-librispeech) - but I am encountering some difficulties, specifically I'm quite sure the EncDecFineTune class is incorrect and where the ctc_cost hparams is defined. 

My steps so far:
1) import the pretrained model from the speechbrain repos
2) Created the .json from my data to use during fine tune as per the pretraining/finetuning tutorial
3) Modified the EncDecFineTune class (from the tutorial) taking the relevant parts from the transformer [train.py] ([weblink](https://github.com/speechbrain/speechbrain/blob/develop/recipes/LibriSpeech/ASR/transformer/train.py))  following the tutorial
4) Created a 'modules' dictionary containing modules of the model I wish to flag as gradients required in the on_stage_start method.
6) Created 'hparams' dictionary

The current configuration is running but generating nan loss. 

```
asr_model1 = EncoderDecoderASR.from_hparams(source="speechbrain/asr-transformer-transformerlm-librispeech", savedir="pretrained_models/asr-transformer-transformerlm-librispeech")

Setting up the data pipeline
import speechbrain as sb
import torch
from parse_data import parse_to_json # parse_data is a local library downloaded before (see Installing Dependencies step) 
parse_to_json("./LibriSpeech/dev-clean-2")
from speechbrain.dataio.dataset import DynamicItemDataset
dataset = DynamicItemDataset.from_json("data.json")
dataset = dataset.filtered_sorted(sort_key="length", select_n=100)
# we limit the dataset to 100 utterances to keep the trainin short in this Colab example
dataset.add_dynamic_item(sb.dataio.dataio.read_audio, takes="file_path", provides="signal")

# 3. Define text pipeline:
@sb.utils.data_pipeline.takes("words")
@sb.utils.data_pipeline.provides(
        "words", "tokens_list", "tokens_bos", "tokens_eos", "tokens")
def text_pipeline(words):
      yield words
      tokens_list = asr_model.tokenizer.encode_as_ids(words)
      yield tokens_list
      tokens_bos = torch.LongTensor([asr_model.hparams.bos_index] + (tokens_list))
      yield tokens_bos
      tokens_eos = torch.LongTensor(tokens_list + [asr_model.hparams.eos_index]) # we use same eos and bos indexes as in pretrained model
      yield tokens_eos
      tokens = torch.LongTensor(tokens_list)
      yield tokens

dataset.add_dynamic_item(text_pipeline)
dataset.set_output_keys(["id", "signal", "words", "tokens_list", "tokens_bos", "tokens_eos", "tokens"])

Fine-Tuning the ASR model

from speechbrain.lobes.features import Fbank
import torch 

# Define fine-tuning procedure 
class EncDecFineTune1(sb.Brain):

    def on_stage_start(self, stage, epoch):
        # enable grad for all modules we want to fine-tune
        if stage == sb.Stage.TRAIN:
            for module in [self.modules.CNN, self.modules.Transformer, self.modules.seq_lin,self.modules.ctc_lin]:
                for p in module.parameters():
                    p.requires_grad = True
     
    def compute_forward(self, batch, stage):

        """Forward computations from the waveform batches to the output probabilities."""
        batch = batch.to(self.device)
        wavs, wav_lens = batch.signal
        tokens_bos, _ = batch.tokens_bos

        # compute features
        feats = self.modules.compute_features(wavs)
        feats = self.modules.normalize(feats, wav_lens)

        # forward modules
        src = self.modules.CNN(feats)
        enc_out, pred = self.modules.Transformer(
            src, tokens_bos, wav_lens, pad_idx=0
        )

        # output layer for ctc log-probabilities
        logits = self.modules.ctc_lin(enc_out)
        p_ctc = self.hparams.log_softmax(logits)

        # output layer for seq2seq log-probabilities
        pred = self.modules.seq_lin(pred)
        p_seq = self.hparams.log_softmax(pred)
        hyps = None
        return p_ctc, p_seq, wav_lens,hyps

    def compute_objectives(self, predictions, batch, stage):
        """Computes the loss (CTC+NLL) given predictions and targets."""

        (p_ctc, p_seq, wav_lens,hyps,) = predictions

        ids = batch.id
        tokens_eos, tokens_eos_lens = batch.tokens_eos
        tokens, tokens_lens = batch.tokens

        loss_seq = self.hparams.seq_cost(
            p_seq, tokens_eos,tokens_eos_lens
        )
        self.hparams.ctc_weight = 1 # attempt to force the code to run 
        loss_ctc = self.hparams.ctc_cost(p_ctc, tokens, wav_lens, tokens_lens)
        loss = (
            self.hparams.ctc_weight * loss_ctc
            + (1 - self.hparams.ctc_weight) * loss_seq
        )
        

        return loss

    def fit_batch(self, batch):
        """Train the parameters given a single batch in input"""
        predictions = self.compute_forward(batch, sb.Stage.TRAIN)
        loss = self.compute_objectives(predictions, batch, sb.Stage.TRAIN)
        loss.backward()
        if self.check_gradients(loss):
            self.optimizer.step()
        self.optimizer.zero_grad()
        return loss.detach()

modules1 = {"CNN": asr_model1.mods.encoder.cnn, 
           "Transformer": asr_model1.hparams.Transformer,
           "compute_features": asr_model1.mods.encoder.compute_features, # we use the same features 
           "normalize": asr_model1.mods.encoder.normalize,
           "seq_lin": asr_model1.hparams.seq_lin,
            "ctc_lin": asr_model1.hparams.ctc_lin,
          }
 
hparams1 = {"seq_cost": lambda x, y, z: speechbrain.nnet.losses.nll_loss(x, y, z, label_smoothing = 0.1)
            "ctc_cost": lambda x,y,z,z1: speechbrain.nnet.losses.ctc_loss(x,y,z,z1,blank_index=1), **#blank_index as a dummy variable since we are unsure what to pass**
            "log_softmax": speechbrain.nnet.activations.Softmax(apply_log=True)}

brain = EncDecFineTune1(modules1, hparams=hparams1, opt_class=lambda x: torch.optim.SGD(x, 1e-5))
brain.tokenizer = asr_model1.tokenizer

brain.fit(range(2), train_set=dataset, 
          train_loader_kwargs={"batch_size": 8, "drop_last":True, "shuffle": False})`
```
Any assistance would be greatly appreciated. 

Thanks in advance.

**Comment by TParcollet** [comment]
For finetuning, I would recommend using the Pretrainer instead of Interfaces. Usually, for interfaces, gradients are blocked. If you are using exactly the same model without modification appart from the dataset, why not just using the same recipe, download the output dir from the google drive, and then 1. do the validation to ensure that everything works properly so you get the score than we do; 2. Modify the yaml with your own data ; Otherwise, if you want to use your own train.py with external pretrained model without relying on the original recipe, please consider trying with the Pretrainer instead.

**Comment by jclarke98** [comment]
Thanks a lot for the clear explanation. My issues is now resolved.
